charms/istio-pilot/lib/charms/istio_pilot/v0/ingress.py

```diff
@@ -0,0 +1,189 @@
+    def is_available(self):
+        return self.charm.unit.is_leader() and self.status is None
+
+    @property
+    def is_ready(self):
+        return self.is_available and self.url
+
+    def _check_provider(self, event):
+        if self.is_ready:
+            self.on.ready.emit()
+        elif self.is_available:
+            self.on.available.emit()
+        elif isinstance(self.status, BlockedStatus):
+            self.on.failed.emit()
+
+    def request(
+        self,
+        *,
+        port: int,
+        service: str = None,
+        prefix: str = None,
+        rewrite: str = None,
+        namespace: str = None,
+        per_unit_routes: bool = False,
+    ):
+        """Request ingress to a service.
+
+        Args:
+            service: the name of the target K8s service to route to; defaults to the
+                charm's automatically created service (i.e., the application name)
+            port: the port of the service (required)
+            prefix: the path used to match this service for requests to the gateway;
+                must not conflict with other services; defaults to f"/{service}/"
+            rewrite: the path on the target service to map the request to; defaults
+                to "/"
+            namespace: the namespace the service is in; default to the current model
+            per_unit_routes: whether or not to create URLs which map to specific units;
+                the URLs will have their own prefix of f"{prefix}-unit-{unit_num}" (with
+                tailing slashes handled appropriately)
+        """
+        if not self.charm.unit.is_leader():
+            raise RequestFailed(
+                WaitingStatus(f"Only leader can request ingress: {self.relation_endpoint}")
+            )
+        if self.status is not None:
+            raise RequestFailed(self.status)
+        ingress = get_interface(self.charm, self.relation_endpoint)
+        ingress.send_data(
+            {
+                "namespace": namespace or self.model.name,
+                "prefix": prefix or f"/{self.charm.app.name}/",
+                "rewrite": rewrite or "/",
+                "service": service or self.charm.app.name,
+                "port": port,
+                "per_unit_routes": per_unit_routes,
+            },
+        )
+
+    @cached_property
+    def url(self):
+        """The full ingress URL to reach the target service by.
+
+        May return None if the URL isn't available yet.
+        """
+        try:
+            ingress = get_interface(self.charm, self.relation_endpoint)
+            if not ingress:
+                return None
+        except (NoCompatibleVersions, NoVersionsListed):
+            return None
+        all_data = ingress.get_data()
+        for (rel, app), data in all_data.items():
+            if app is self.charm.app:
+                continue
+            return data["url"]
+        else:
+            return None
+
+    @cached_property
+    def unit_urls(self):
```

Comment:
I would raise if per unit routing is not configured.

Reply:
That seems unnecessarily aggressive.  If you didn't request per-unit routes, you won't get any, and I don't really see that as an error.  Additionally, checking that would require reading the app-level relation data sent by this application which is restricted to only the leader by Juju which would mean non-leader units could never read their own URL. (There is [an open bug against SDI](https://github.com/canonical/serialized-data-interface/issues/27) which also imposes this limitation, but I intend to submit a fix for that.)

Reply:
Writing is restricted to the leader, not reading :-)

Reply:
Reading your own app-level relation data is, in fact, restricted to the leader by Juju. I don't agree with that, and maybe it will change at some point, but AFAIK that is still a limitation.

Reply:
I was a little unclear about the SDI bug; that *also* prevents non-leaders from reading *remote* app-level relation data, which is, in fact, allowed by Juju.